first we import all the neccessary packages for working with the data

In [1]:
import os
from IPython.display import display
#import hoboreader
import pandas as pd
import numpy as np
import shutil
import matplotlib as mp
import re

next we ensure that our working directory is set correctly so that when we're working with file paths they always work relative to our project

In [2]:
#os.chdir("..")
os.chdir("C:\\Users\\samer\\OneDrive - Cal Poly\\Classes\\SPR_Fellowship\\SPR_Fellowship")
os.getcwd()

'C:\\Users\\samer\\OneDrive - Cal Poly\\Classes\\SPR_Fellowship\\SPR_Fellowship'

First well create a function for searching through complex fill structures and returning the paths to files of only a certain file type

In [3]:
def search_filetype(path = ".", extentension = ""):
    """
    :param path: string, where to search
    :param extentension: string, what files to look for
    :return: list, paths to all files with a matching extension
    """
    paths = []
    for dirpath, dirnames, filenames in os.walk(path):
        for filename in filenames:
            if filename[-len(extentension):] == extentension:
                paths.append(dirpath + "/" + filename)
    return paths

Well also want some test data to work with along the way before we start working with thousands of files

In [4]:
list_ex_headers = search_filetype(path="LegacyFiles/All_txt/txt_header_examples/", extentension=".txt")

print(list_ex_headers)

['LegacyFiles/All_txt/txt_header_examples//Al RRG_030916.txt', 'LegacyFiles/All_txt/txt_header_examples//AL RRG_060403a.txt', 'LegacyFiles/All_txt/txt_header_examples//AL_RRG_091222 remove first tip.txt']


# Dealing with Headers
Next we create a function for working with the messsy headers created by the text files.  There are a few different ones to worry about, we have to manage for each one, while preserving inprotant information, like file name and serial numbers.

In [5]:
def header_fixer(path = "."):
    """
    :param path: path to the txt/csv file
    :return: a dataframe with a fixed header
    """
    csv = pd.read_csv(path, sep="\t") #read in the csv just to check its structure
    
    if csv.shape[1] < 2:  #because files with SN have a multiline header this will filter those files out
        serial_Num = re.findall("[0-9]{5,6}", csv.columns[0])
        csv = pd.read_csv(path, header=1, sep="\t")
        csv["Serial Number"] = serial_Num[0]
    else:
        csv = pd.read_csv(path, sep="\t")
        csv["Serial Number"] =  None

    return csv


great, now lets test it out on our small sample of txt files we created before

In [6]:

ex_headers_csv_list = []

for path in list_ex_headers:
    csv = header_fixer(path = path)
    csv["FileName"] = path.split(sep="/")[-1]
    ex_headers_csv_list.append(csv) #create a list of test dataframes
    display(csv.head())

Date Time  Event (Rain Bucket Tip) Serial Number  \
0  06/18/03 08:40:17.0                     0.00          None   
1  09/09/03 14:19:42.0                     0.01          None   
2  09/09/03 15:43:36.5                     0.02          None   
3  09/09/03 16:33:22.0                     0.03          None   
4  09/09/03 17:20:45.5                     0.04          None   

            FileName  
0  Al RRG_030916.txt  
1  Al RRG_030916.txt  
2  Al RRG_030916.txt  
3  Al RRG_030916.txt  
4  Al RRG_030916.txt

Date Time  Event (Rain bucket tip) Serial Number  \
0  03/26/06 17:58:38.0                     0.00        531980   
1  03/27/06 13:23:29.5                     0.01        531980   
2  03/27/06 13:37:13.5                     0.02        531980   
3  03/27/06 13:47:53.5                     0.03        531980   
4  03/27/06 13:53:57.5                     0.04        531980   

             FileName  
0  AL RRG_060403a.txt  
1  AL RRG_060403a.txt  
2  AL RRG_060403a.txt  
3  AL RRG_060403a.txt  
4  AL RRG_060403a.txt

Date Time  Event (Events) Serial Number  \
0  12/15/09 16:31:03.0        0.000000          None   
1  12/15/09 16:31:24.0        0.010870          None   
2  12/16/09 09:04:57.0        0.021739          None   
3  12/16/09 09:20:12.5        0.032609          None   
4  12/16/09 09:40:41.0        0.043478          None   

                             FileName  
0  AL_RRG_091222 remove first tip.txt  
1  AL_RRG_091222 remove first tip.txt  
2  AL_RRG_091222 remove first tip.txt  
3  AL_RRG_091222 remove first tip.txt  
4  AL_RRG_091222 remove first tip.txt

# deal with column names
That seems to be working okay, but there are definitely some differences in column names that are going to give us issues later, so lets create a function to deal with those. 

In [7]:
def columnNames_handler(dataframe, column_names = []):
    """takes in a pandas dataframe, and replaces the column names with 
    the ones provided in the column_names list.  if lengths don't match, it retains 
    names exceeding the list, or discards names from the list if there are less 
    columns than the length of the list.

    Args:
        dataframe (pandas dataframe): a pandas dataframe with a cleaned header.
        column_names (list, optional): list of column names (string). Defaults to [].
    """
    df_num_cols = len(dataframe.columns)

    if len(column_names) > df_num_cols: #if there are too many column names given
        column_names = column_names[:(df_num_cols)] 
    elif len(column_names) < df_num_cols: #if there are too few column names given
        column_names = column_names + dataframe.columns[len(column_names):]
    
    dataframe.columns = column_names

    return dataframe



and a test

In [8]:
index = 0

column_names = ["DateTime","BucketTip_Cumm","Serial Number","FileName"]
for df in ex_headers_csv_list:
    ex_headers_csv_list[index] = columnNames_handler(dataframe=df, column_names=column_names)
    print(df.columns)
    index += 1

Index(['DateTime', 'BucketTip_Cumm', 'Serial Number', 'FileName'], dtype='object')
Index(['DateTime', 'BucketTip_Cumm', 'Serial Number', 'FileName'], dtype='object')
Index(['DateTime', 'BucketTip_Cumm', 'Serial Number', 'FileName'], dtype='object')


# deal with cummulative values
As you can see, individual bucket tips were summed up progressively as more tips were recorded.  This is helpful for finding the amount of total rainfall for the length of time between log pulls from precipitation monitors, but will make no sense when combined with other datasets.  There are a couple options for dealing with this:

1. just make one record for each timespan between log pulls in the final dataset
2. solve for each individual bucket tip and make them non-cummulative 

option 2 will allow us to preserve the data in its most granular form, so that future analyists can aggrigate the data as they see fit, so that is what we will work towards.
https://stackoverflow.com/questions/27581942/how-can-i-get-back-real-value-from-cumulative-sum-in-pandas

In [9]:
ex_headers_csv_list[0]["BucketTip_Cumm"] #check accumulation of values

0     0.00
1     0.01
2     0.02
3     0.03
4     0.04
5     0.05
6     0.06
7     0.07
8     0.08
9     0.09
10    0.09
Name: BucketTip_Cumm, dtype: float64

In [10]:
def cumm_to_individual(dataframe, column_index_or_name, indiv_column_name):
    """given a column with cummulative values, add a column with individual non-cummulative records

    Args:
        dataframe (pandas df): _description_
        column_index_or_name (int or str): _description_
        column_name (str): _description_
    """
    if type(column_index_or_name) == int:
        col = dataframe.columns[column_index_or_name]
    else:
        col = column_index_or_name
        if col not in dataframe.columns:
            return dataframe #if the specified column doesnt exist, just return the dataframe
    dataframe[indiv_column_name] = dataframe[col].diff().fillna(dataframe[col].iloc[0])
    return dataframe

cumm_to_individual(dataframe=ex_headers_csv_list[0], column_index_or_name="BucketTip_Cumm", indiv_column_name="precip_individual")["precip_individual"]

0     0.00
1     0.01
2     0.01
3     0.01
4     0.01
5     0.01
6     0.01
7     0.01
8     0.01
9     0.01
10    0.00
Name: precip_individual, dtype: float64

looks much better, now just iterate over all the datasets to give them all that value.

# list all files
Know we need to make a list of all the files we want to try to aggregate.

In [11]:
def search_filetype(path = ".", extentension = ""):
    """
    :param path: string, where to search
    :param extentension: string, what files to look for
    :return: list, paths to all files with a matching extension
    """
    paths = []
    for dirpath, dirnames, filenames in os.walk(path):
        for filename in filenames:
            if filename[-len(extentension):] == extentension:
                paths.append(dirpath + "/" + filename)
    return paths

#list of all text files in the legacy files directory
list_text_paths = search_filetype(path = "LegacyFiles/", extentension=".txt") 

# Put it all together
Now we just iterate through all of them and see how it looks!

In [38]:
df_list = []
ticker = 0
error_files = []

for text_path in list_text_paths:
    try:
        #print(ticker)
        ticker += 1
        df = header_fixer(text_path) #read in each txt file using the header_fixer function
        df = columnNames_handler(df, column_names=column_names)
        df["FileName"] = text_path.split(sep="/")[-1]
        df = cumm_to_individual(df, column_index_or_name="BucketTip_Cumm", indiv_column_name="precip_individual")
        if len(df.columns)<=5:
            df_list.append(df)
    except: 
        error_files.append(text_path)


In [39]:
print(f"number of total txt: {len(list_text_paths)} \nnumber of error files: {len(error_files)}")

number of total txt: 1724 
number of error files: 86


We'll create a directory to save the error files, but 86 out of 1724 isnt too bad. After we run through the all the .csv files then we can get on directory full of only files that couldn't be processed.  They may need to be looked at manually depending on the total number of files.

In [20]:
dest_txt_errors = os.getcwd()+"/LegacyFiles/TxtErrorFiles/"
os.makedirs(dest_txt_errors)
for error_file in error_files:
    shutil.copy2(error_file, dest_txt_errors) #copy all error files into the new directory


Then we just create a full dataset of all the text files we just processed.  Here we also filter out large precip numbers that don't make sense at our known bucket tip scale of .001", and we convert them by the scale factor to match the rest of the data.

In [34]:
full_df = pd.concat(df_list, names=column_names, ignore_index=True)
full_df["precip_individual"] = np.where((full_df.precip_individual >= 1), full_df.precip_individual*0.001, full_df.precip_individual)

display(df[0:20])


DateTime  BucketTip_Cumm Serial Number  \
0   12/15/09 16:31:03.0        0.000000          None   
1   12/15/09 16:31:24.0        0.010870          None   
2   12/16/09 09:04:57.0        0.021739          None   
3   12/16/09 09:20:12.5        0.032609          None   
4   12/16/09 09:40:41.0        0.043478          None   
5   12/16/09 10:40:35.0        0.054348          None   
6   12/16/09 10:57:48.0        0.065218          None   
7   12/16/09 11:32:49.5        0.076087          None   
8   12/16/09 12:40:00.5        0.086957          None   
9   12/17/09 03:26:02.5        0.097826          None   
10  12/21/09 08:11:05.5        0.108696          None   
11  12/21/09 11:26:27.5        0.119566          None   
12  12/21/09 12:05:51.0        0.130435          None   
13  12/21/09 12:11:00.5        0.141305          None   
14  12/21/09 12:17:02.5        0.152174          None   
15  12/21/09 12:25:05.5        0.163044          None   
16  12/21/09 12:39:22.0        0.173914          None   
17  12/21/09 13:07:34.5        0.184783          None   
18  12/21/09 13:11:35.5        0.195653          None   
19  12/21/09 13:15:46.5        0.206522          None   

                              FileName  precip_individual  
0   AL_RRG_091222 remove first tip.txt           0.000000  
1   AL_RRG_091222 remove first tip.txt           0.010870  
2   AL_RRG_091222 remove first tip.txt           0.010870  
3   AL_RRG_091222 remove first tip.txt           0.010870  
4   AL_RRG_091222 remove first tip.txt           0.010870  
5   AL_RRG_091222 remove first tip.txt           0.010870  
6   AL_RRG_091222 remove first tip.txt           0.010870  
7   AL_RRG_091222 remove first tip.txt           0.010870  
8   AL_RRG_091222 remove first tip.txt           0.010870  
9   AL_RRG_091222 remove first tip.txt           0.010870  
10  AL_RRG_091222 remove first tip.txt           0.010870  
11  AL_RRG_091222 remove first tip.txt           0.010870  
12  AL_RRG_091222 remove first tip.txt           0.010869  
13  AL_RRG_091222 remove first tip.txt           0.010870  
14  AL_RRG_091222 remove first tip.txt           0.010869  
15  AL_RRG_091222 remove first tip.txt           0.010870  
16  AL_RRG_091222 remove first tip.txt           0.010870  
17  AL_RRG_091222 remove first tip.txt           0.010869  
18  AL_RRG_091222 remove first tip.txt           0.010870  
19  AL_RRG_091222 remove first tip.txt           0.010869

then to keep things organized we just write this intermediate file to a directory in our project.

In [37]:
txt_csv_write_dst = os.getcwd() + "/csv_from_legacy/All_txt_as_csv/txts_to_csv.csv"
if not os.path.exists(txt_csv_write_dst): #only do this once, don't repeat on additional script runs
    full_df.to_csv(txt_csv_write_dst)

## repeat for csv files from legacy exports
now we just repeat the process for all of the .csv files exported from .hobo and .dtf files created directly by the hobo loggers. starting with .dtfs

In [43]:
dtf_csv_file_list = search_filetype(path="csv_from_legacy/All_dtf_as_csv/", extentension=".csv")
f"number of .csv from .dtf: {len(dtf_csv_file_list)}"

'number of .csv from .dtf: 3202'

The dataframe structure from these is a bit different though

In [50]:
display(pd.read_csv(dtf_csv_file_list[0]))

Plot Title: AL-RRG
#  Date Time, GMT-08:00    Rain Bu (0.009999999776482582), Units (LGR S/N...
1  01/13/04 08:58:15 AM                                             28025.11
2  01/13/04 09:01:57.5 AM                                           28025.12
3  01/13/04 09:02:14.5 AM                                           28025.13
4  01/13/04 09:02:32 AM                                             28025.14
...                                                                      ...
94 01/13/04 10:10:05.5 AM                                           28026.04
95 01/13/04 10:11:32.5 AM                                           28026.05
96 01/13/04 10:13:20 AM                                             28026.06
97 01/13/04 10:16:11 AM                                             28026.07
98 01/13/04 10:25:19.5 AM                                           28026.07

[99 rows x 1 columns]

a couple things to note here:
 - times are listed with AM or PM at the End - we'll need to change this to international time for a standard
 - plot Title is listed in the first column, first cell - we'll want to save that to its own column
 - S/N is embeded in the second column header... what a mess...
 - Calibration for units is also embeded in the second column header... who designed this??
 - Like our previous dataset bucket tips are recorded cummulative, but here they are lifetime cummulative, this will change how we deal with it slightly


### dealing with dates
we'll start by parsing out the date, so we can convert the time to international.  We'll create a function for this since were going to have to retroactively apply this to our previous .txt dataset too so we can have a column for date, and a seperate column for time.

In [52]:
test_dtf_files = dtf_csv_file_list[0:2]

